File: ts/explosion.ts
import { TILE, ExplosionSize } from './constants.js';
import { IMAGES } from './images.js';

export class Explosion {
    x: number;
    y: number;
    size: ExplosionSize;
    frame: number;
    maxFrame: number;
    alive: boolean;
    timer: number;

    constructor(x: number, y: number, size: ExplosionSize = 'normal') {
        this.x = x;
        this.y = y;
        this.size = size;
        this.frame = 0;
        this.maxFrame = 8;
        this.alive = true;
        this.timer = 0;
    }

    update(): void {
        this.timer++;
        if (this.timer > 4) {
            this.timer = 0;
            this.frame++;
            if (this.frame >= this.maxFrame) this.alive = false;
        }
    }

    draw(ctx: CanvasRenderingContext2D): void {
        if (!this.alive) return;
        const img: HTMLImageElement | undefined = IMAGES.explosion;
        const scale: number = this.size === 'big' ? 2 : (this.size === 'small' ? 0.6 : 1);
        const progress: number = this.frame / this.maxFrame;
        const alpha: number = 1 - progress;
        const drawSize: number = TILE * scale;

        if (img && this.frame < 3) {
            ctx.globalAlpha = alpha;
            ctx.drawImage(img, this.x - (scale - 1) * TILE / 2, this.y - (scale - 1) * TILE / 2, drawSize, drawSize);
            ctx.globalAlpha = 1;
        } else {
            const cx: number = this.x + TILE / 2;
            const cy: number = this.y + TILE / 2;
            const radius: number = (8 + this.frame * 4) * scale;
            ctx.fillStyle = `rgba(255, ${120 + this.frame * 15}, 0, ${alpha})`;
            ctx.beginPath();
            ctx.arc(cx, cy, radius, 0, Math.PI * 2);
            ctx.fill();
            ctx.strokeStyle = `rgba(255, 200, 0, ${alpha * 0.5})`;
            ctx.lineWidth = 2;
            ctx.beginPath();
            ctx.arc(cx, cy, radius + 4, 0, Math.PI * 2);
            ctx.stroke();
        }
    }
}

File: explosion.py
from constants import *
from pgzero.builtins import sounds, images
import pygame


class Explosion:
    def __init__(self, x, y, size='normal', obj_size=TILE):
        self.x = x
        self.y = y
        self.size = size
        self.obj_size = obj_size
        self.frame = 0
        self.max_frame = 8
        self.alive = True
        self.timer = 0
        if size == 'big':
            sounds.enemy_explode.play()
        elif size == 'small':
            sounds.bullet_hit_wall.play()
        elif size == 'destroy':
            sounds.player_explode.play()
        elif size == 'brick':
            sounds.destory_wall.play()

    def update(self):
        self.timer += 1
        if self.timer > 4:
            self.timer = 0
            self.frame += 1
            if self.frame >= self.max_frame:
                self.alive = False

    def draw(self, screen):
        if not self.alive:
            return
        scale = 1 if self.size == 'destroy' else (1 if self.size == 'big' else (0.6 if self.size in ('small', 'brick') else 1))
        progress = self.frame / self.max_frame
        img = images.explosion
        draw_size = int(TANK_SIZE * scale * (0.5 + progress * 0.5))
        if draw_size < 4:
            return
        scaled = pygame.transform.scale(img, (draw_size, draw_size))
        scaled.set_alpha(int(255 * (1 - progress * 0.7)))
        cx = self.x + self.obj_size / 2
        cy = self.y + self.obj_size / 2
        ox = int(cx - draw_size / 2)
        oy = int(cy - draw_size / 2)
        screen.blit(scaled, (ox, oy))
